### ISSUE TITLE
Allow specifying a custom `initialTime` for ReactiveDatePicker

### ISSUE THREAD

**Issue by ezet** [joanpablo/reactive_forms#264]
I have a use case where the `firstDate` depends on the current date. The property is stored in a database, but is editable. This means that I end up with a case where a user selected a date that was valid at the time they selected it. However, if the then selected date has passed, the user is no longer able to edit the date, to set a new, valid date. 

The specific error is something like:
`'package:flutter/src/material/date_picker.dart': Failed assertion: line 167 pos 5: '!initialDate.isBefore(firstDate)': initialDate 2022-01-26 00:00:00.000 must be on or after firstDate 2022-01-27 00:00:00.000.`

The issue is obvious, the date is set to a past date, which is no longer valid. But the user should still be able to edit it. Currently, the only solution is to modify the `Form` data in order to update the value to a valid date, before opening the picker. However this is causes a whole range of other issues and problems.

It seems that an obvious and simple fix would be to let us specify a custom nullable `initialDate`, which would take precedence over any other date when provided.

**Comment by ezet** [comment]
Another similar case that would be solved by the same fix is when the `firstDate > now`, but the form value is null. `initialDate` will then default to `now`, which fails validation and the same error occurs. And I don't want to modify the form data unless the user actually selects a value. It should remain null if the user doesn't touch the field.

**Comment by ezet** [comment]
Another, and perhaps better solution, would be to consider the `firstDate` in the `_getInitialTime()` function, and default to `firstDate` if the fieldValue is null and firstDate is specified. And also default to firstDate if fieldValue < firstdate

**Comment by vasilich6107** [comment]
Hi @ezet 
As you posted in your first message - the issue comes from “ package:flutter/src/material/date_picker.dart” which is behind our control as far as we just proxy values to the component.
Is you have any custom logic that does not fit into standard flutter control you can create your own component referring to exiting date picker as an example.

**Comment by ezet** [comment]
@vasilich6107 Yes, the error is thrown from a framework package, but the problem arises because this library passes an invalid argument to the framework "date_picker", without giving us a way to change that behavior. The framework picker accepts an argument `initialDate`, which is invalid in this case. ReactiveDatePicker sets that value through it's own calculation, which has buggy behavior, as I described in detail above. There are two solutions I can think of, as mentioned:
1. Add a new argument on ReactiveDatePicker, that lets me override the `initialDate`
2. Fix the bugged calculation, so that it doesn't provide an invalid value to the framework picker. https://github.com/joanpablo/reactive_forms/issues/264#issuecomment-1022579701 describes a possible fix.

If required, I can provide a PR, but I'd first need to know whether this is something you'd accept, and which solution is preferred.

**Comment by vasilich6107** [comment]
@ezet I suppose that giving an ability to override `_getInitialTime` would be a universal solution

**Comment by vasilich6107** [comment]
@ezet i'm maintaining the package `reactive_datetime_picker`
https://github.com/artflutter/reactive_forms_widgets/blob/master/packages/reactive_date_time_picker/lib/reactive_date_time_picker.dart

I will try to add an ability to override `_getInitialTime` which could be helpfull
It will be the shortest path cause the releases of `reacitve_forms` are unpredictable)
There are 7 PR from myself waiting to be merged for month already

**Comment by ezet** [comment]
@vasilich6107 Thank you, this should suffice for now! I realize this should've been reported as an issue in your repo, not here. But thanks for assisting anyways! You can consider also applying the logic correction I proposed in my comment, while also allowing callers to provide a custom value. I don't think my proposal would be breaking, as it would only use a different `initialDate` in cases that would otherwise result in an exception.

**Comment by vasilich6107** [comment]
Technically it was easier to me to give an ability to override the initial date getter)

**Comment by vasilich6107** [comment]
@ezet I will try to find time to apply the same improvement here...
But only the gods know when it will be merged

**Comment by vasilich6107** [comment]
@ezet i'm also working on generator plugin for reactive forms
https://github.com/artflutter/reactive_forms_generator

It would be great if you could try it

**Comment by joanpablo** [comment]
Hi @ezet,

Thanks for the issue, I guess the solution is to fix the "bug" in the initial date calculation because you already set the value of the date in the Control. 

Or maybe both, if you provide the initialDate to the Widget and the value of Control is Null then the control may assume the value of the Widget. (But I will work on fix the calculation first)

I will see what is the error and will try to release the fix asap.

@vasilich6107 sorry for the delay on PR. I know they are almost all related to FocusNode, I haven't prioritized them yet since they are not a critical issues, but you are right about the delay.

**Comment by joanpablo** [comment]
Hi again @ezet,

Do you think the argument of **initialDate** should be a method or just another optional property like **firstDate**, and **lastDate**? If we add an extra argument to the **ReactiveDatePicker** would that be enough for you? (Of course I will reimplement the calculation of the _getInitialDate as a separated fix)

**Comment by joanpablo** [comment]
Hi @ezet and @vasilich6107,

this is the New implementation of the **_getInitialDate**

```dart
static DateTime _getInitialDate(DateTime firstDate, DateTime lastDate, DateTime? defaultInitDate) {
    var initialDate = defaultInitDate ?? DateTime.now();

    if (initialDate.compareTo(firstDate) < 0) {
      initialDate = firstDate;
    } else if (initialDate.compareTo(lastDate) > 0) {
      initialDate = lastDate;
    }

    return initialDate;
  }
```

this is how I call it

```dart
initialDate: _getInitialDate(firstDate, lastDate, initialDate ?? field.value)
```

As you can see, now the **ReactiveDatePicker** will have an optional new argument: **initialDate**, this one will have more priority than the **field.value**. If both values are null then the **DateTime.now()** will be used. And inside the method I make sure that the value is inside the range of **firstDate** and **lastDate**.

That should fix the issue.

I will add some tests and let you know when it is released.

**Comment by joanpablo** [comment]
Hi @ezet and @vasilich6107,

I have released a new version v11.0.1 of **Reactive Forms** which not only includes an optional argument **initialDate**, but also fix the initialDate calculation.

Please if you still have issues, don't hesitate to reopen the issue or create a new one.

**Comment by ezet** [comment]
> Do you think the argument of **initialDate** should be a method or just another optional property like **firstDate**, and **lastDate**? If we add an extra argument to the **ReactiveDatePicker** would that be enough for you? (Of course I will reimplement the calculation of the _getInitialDate as a separated fix)
